lib/kinesis_client/stream/shard/producer.ex

```diff
@@ -25,8 +33,7 @@ defmodule KinesisClient.Stream.Shard.Producer do
     :app_name,
     :app_state_opts,
     :lease_owner,
-    :shard_closed_timer,
-    shutdown_delay: 300_000,
```

Comment:
Big-ish change: kcl_ex used to have a 5 minute delay before stopping the shard. I've gotten rid of that. We may need to revisit this so that we have some mechanism to figure out when the downstream Broadway pipeline has fully drained? Or do we think relying on Elixir supervision magic is enough to signal to broadway that the pipeline is coming down and it needs to flush its work?

This should hold merging until we've got an answer.

Reply:
Turns out elixir supervision magic is the answer here: https://hexdocs.pm/broadway/architecture.html#graceful-shutdowns

Reply:
According to the [`Broadway.start_link/1` docs](https://hexdocs.pm/broadway/Broadway.html#start_link/2-options), pipeline config can take a `:shutdown` option documented as:

> `:shutdown` ([pos_integer/0](https://hexdocs.pm/elixir/typespecs.html#basic-types)) - Optional. The time in milliseconds given for Broadway to gracefully shutdown without discarding events. The default value is 30000.

So, while we do get a graceful shutdown window built into Broadway, it's still time bound. 

Do we think 30s is enough time? I think it likely is in vast majority of cases. But, I could foresee edges where it's not. However, I believe we're using the 30s default right now in production for normal shutdowns (i.e. deployment rollout), and we haven't ever noticed issues. Still, it concerns me that we don't have full insight into this.

So I'm thinking we either A) configure this `:shutdown` option to 300s to maintain previous config for splits, or B) log a ticket to return here and instrument shutdown drains for our pipelines to get better insight into data loss etc.

What do you think?

Reply:
Agree. Let's configure it to use 300 seconds. That's the max limit, so once the pipeline is fully drained it should shutdown and move on anyway, which is the behavior that we want.

The secondary question is whether we wait for the pipeline to drain before firing up any child shards? I think that's the old behavior here, and we're not really retaining it since we tell the shard to shutdown but I think that is an asynchronous `cast` internally - we don't wait for the shutdown to be complete before appending the child shards.

One way to handle that would be to tell the coordinator that the parent shard is complete, and to give the child shards that will be booted up. We then monitor the child shard, and then tell it to shut down. We also record in the coordinator's state the child shards that ought to boot up once the parent is shut down. We handle the `DOWN` message from the parent, and use that to then start up the child shards.

Thoughts?

Reply:
> Let's configure it to use 300 seconds. That's the max limit, so once the pipeline is fully drained it should shutdown and move on anyway, which is the behavior that we want.

👍 

> The secondary question is whether we wait for the pipeline to drain before firing up any child shards?

Ah right, I hadn't even thought of this. 

I think we do need to do this, since it's necessary to ensure consumers handle messages in a deterministic/ordered manner.

> One way to handle that would be to tell the coordinator that the parent shard is complete, and to give the child shards that will be booted up. We then monitor the child shard, and then tell it to shut down. We also record in the coordinator's state the child shards that ought to boot up once the parent is shut down. We handle the DOWN message from the parent, and use that to then start up the child shards.

I believe this makes sense to me!

Reply:
Closing this out, and we are using 300 seconds for the Broadway shutdown timeout (look for the `:timer.minutes(5)` in the changes).

I've also updated how we handle the shutdown process, so we use `Process.monitor` to keep an eye out for the process actually shutting down. So, the process goes something like this:

1. Producer hits the end of the stream. Messages are forwarded via Broadway/Genstage, and are fully consumed in some indeterminate time frame.
2. Producer tells the Coordinator it's done via a `GenServer.call` and which children will continue the stream, then marks itself internally as `closed` so it basically stops doing any work for processors.
3. Coordinator marks the shard as closed in Dynamo, then records that the children shard(s) ought to start after the parent shard(s) terminate. Those children will start if all of its parent shards are closed. If they aren't, it skips out. This is going to be a problem when we introduce more intelligent shard balancing across a cluster of workers (how do other pods know when to try and boot up a shard?).
    - A really easy fix to that would be for us to have a timer that polls Kinesis (semi-randomly to avoid hitting rate limits across a cluster) looking for shards and making sure they are all started. Shards will race to acquire a lease, and then ensuring that those leased shards monitor for changes in the parent shard state. We're not there yet, though.
4. Once the shard is finally down, the coordinator attempts to start the child shards.

We're going to still want to really closely monitor our streams once we get this into something like infra-dev and definitely infra-prod, and be prepared to rollback, as the potential for there to be a mistake in here is still pretty high.

Reply:
> This is going to be a problem when we introduce more intelligent shard balancing across a cluster of workers (how do other pods know when to try and boot up a shard?).

This tradeoff sounds good to me, since we haven't yet prioritized work on shard balancing. It might be some time until we actually invest in that.

Reply:
Right, and I think it's pretty easy to build up a variant of this.